[Sema] Fix and cleanup distributed `id`/`actorSystem` synthesis

hamishknight opened the pull request:
- Ensure `associatedtype` inference never considers `id` and `actorSystem` properties for a distributed actor since doing so would either result in cycles or resolving using `Identifiable`'s default `id`
- Consolidate synthesis logic that was previously duplicated across witness derivation and separate requests to trigger synthesis
- Always synthesize `id` and `actorSystem` for a distributed actor, allowing redeclaration checking to diagnose user-defined versions
- Lazily compute the interface types for `id` and `actorSystem`, ensuring the synthesis logic doesn't kick semantic requests
- Install `id` and `actorSystem` as part of `synthesizeSemanticMembersIfNeeded` to ensure they're always injected into name lookup results (this is the actual fix for rdar://162800185)
- Remove `Identifiable` derivation code and `ImplicitMemberAction::ResolveDistributedActor`, neither of which should be necessary now

rdar://162800185

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

ktoso commented on include/swift/AST/Decl.h:
```diff
@@ -374,6 +374,10 @@ bool conflicting(ASTContext &ctx,
                  bool *wouldConflictInSwift5 = nullptr,
                  bool skipProtocolExtensionCheck = false);
 
+enum class SpecialDistributedProperty {
```
```suggestion
enum class SpecialDistributedActorProperty {
```

Just to that it's clearer that we mean specifically properties of a distributed actor; (a "distributed property" exists and is `distributed var {}` but that's different)

hamishknight committed:
[Sema] Install `id`/`actorSystem` in `synthesizeSemanticMembersIfNeeded`

Now that the synthesis logic for these doesn't involve kicking
semantic requests we can synthesize directly as part of name lookup.
This ensures they get synthesized for secondaries and we don't try
and use Identifiable's default `id`.

rdar://162800185

hamishknight committed:
[Sema] NFC: Inline `id` and `actorSystem` synthesis logic

These can be inlined into their respective requests.

hamishknight committed:
[Sema] Avoid inferring `associatedtype` from special distributed vars

Avoid relying on the cycle detection logic to exclude these and
enforce that we never consider them since computing their interface
type relies on the type witnesses already being resolved. This also
means we can no longer pick up Identifiable's default implementation
`id`, which we never want for a `distributed` actor. This then lets 
us remove the special-cased eager inference logic.

hamishknight committed:
[AST] Avoid returning null Type from `getAssociatedTypeOfDistributedSystemOfActor`

Make sure we produce an error for a broken stdlib and produce ErrorType
instead.

hamishknight committed:
[Sema] Lazily compute type of `id` and `actorSystem` properties

Delay their interface type computation until they are actually
type-checked. This ensures their synthesis logic doesn't produce
cycles.

hamishknight committed:
[Sema] Remove unnecessary logic in `GetDistributedActorSystemPropertyRequest`

This is checking a conformance for something defined in the Distributed
module itself, so things would be very broken if it failed.

hamishknight committed:
[AST] Introduce `isSpecialDistributedProperty`

And use this in `evaluateMembersRequest`.

hamishknight committed:
[Sema] Consolidate distributed actor property synthesis logic

Have the witness derivation logic call into the synthesis requests.

hamishknight committed:
[Sema] Always synthesize `id` and `actorSystem` properties

Only fallback to name lookup for serialized modules and swift
interfaces. This means we can then implement the user-declared
`id`/`actorSystem` error as part of redeclaration checking.

hamishknight committed:
[AST] NFC: Reorder headers in `DistributedDecl.cpp`

hamishknight committed:
[Sema] Remove Identifiable derivation for distributed actors

This shouldn't be necessary anymore since we we'll no longer try to
infer `ID` from Identifiable's default implmentation of `id`.

hamishknight commented:
@swift-ci please test

xedin commented:
Thanks, @hamishknight! I will take a look on Monday.

hamishknight committed:
[Sema] Remove `ImplicitMemberAction::ResolveDistributedActor`

The member names that triggered this don't even match the requirements
in the Distributed module anymore, and
`installDistributedActorIfNecessary` was only called for structs which
can't be distributed actors. Rip it out.

hamishknight commented:
@swift-ci please test source compatibility

ktoso commented on test/Distributed/rdar162800185.swift:
```diff
@@ -0,0 +1,21 @@
+import Distributed
+
+func foo<T: DistributedActor>(_: T.Type) where T.ActorSystem == LocalTestingDistributedActorSystem {}
+func bar() { foo(A.self) }
```
Hmm, this doesn't quite reproduce rdar://162800185; 
This works fine on existing `main` already, so we're missing something here

hamishknight commented:
@swift-ci please smoke test

hamishknight committed:
[AST] Replace a couple more checks for vars named `id`/`actorSystem`

Use `isSpecialDistributedProperty` or the synthesis requests instead.
While here, also make `isSpecialDistributedProperty` agnostic to
interface and serialized module cases.

hamishknight commented:
@swift-ci please test source compatibility

xedin reviewed (approved):
Excellent, thank you!

hamishknight commented on lib/Sema/CodeSynthesisDistributedActor.cpp:
```diff
@@ -944,79 +800,121 @@ FuncDecl *GetDistributedThunkRequest::evaluate(Evaluator &evaluator,
+  // to find this through name lookup.
+  auto *DC = nominal->getDeclContext();
+  if (!DC->getParentSourceFile() || DC->isInSwiftinterface())
+    return lookupDistributedActorProperty(nominal, C.Id_id);
```
AFAICT we already have checks for them

https://github.com/swiftlang/swift/blob/73b03ed41453c1ad9bc8a06ff3404d0c12863f40/test/ModuleInterface/distributed-actor.swift#L35-L36

ktoso reviewed (approved):
Thank you, I also confirmed in an adopter project this solves the issue!
